File: 113_path_sum_II.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> res = new ArrayList<>();
        findAllPaths(res, root, sum, new ArrayList<>());
        return res;
    }
    
    private void findAllPaths(List<List<Integer>> res, TreeNode root,
                              int sum, List<Integer> temp) {
        if (root == null) return;
        
        sum -= root.val;
        temp.add(root.val);
        
        if (root.left == null && root.right == null) {
            if (sum == 0) {
                res.add(new ArrayList<>(temp));   
            }
        }
        
        findAllPaths(res, root.left, sum, temp);
        findAllPaths(res, root.right, sum, temp);
        temp.remove(temp.size() - 1);
    }
}

File: python/113_path_sum_II.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        res = []
        self.findPath(root, sum, res, [])
        return res

    def findPath(self, root: TreeNode, sum: int, paths: List[List[int]], temp: List[int]):
        if not root: return

        sum -= root.val
        temp.append(root.val)

        if not root.left and not root.right:
            if sum == 0:
                paths.append(list(temp))

        self.findPath(root.left, sum, paths, temp)
        self.findPath(root.right, sum, paths, temp)
        temp.pop()
